File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        max_val=0
        st=[-1]
        for i in range(len(s)):
            if s[i] == "(":
                st.append(i)
            else:
                st.pop()
                if not st:
                    st.append(i)
                max_val=max(max_val,i-st[-1])
        if max_val!=0:
            return max_val 
        return 0

File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.cpp
#include<vector>
#include<algorithm>
class Solution {
public:
    int longestValidParentheses(string s) {
        vector<int> stack;
        stack.push_back(-1);
        int max_ans=0;
        for(int i=0;i<s.size();i++)
        {
            if(s[i]=='(')
            {
                stack.push_back(i);
            }
            else
            {
                stack.pop_back();
                if(stack.empty())
                {
                    // cout<<"FD";
                    stack.push_back(i);
                }
                max_ans=max(max_ans,i-stack[stack.size()-1]);
            }
        }
        return max_ans;
    }
};